## Оптимизация выполнения кода, векторизация, Numba

Материалы:
* Макрушин С.В. Лекция 3: Оптимизация выполнения кода, векторизация, Numba
* IPython Cookbook, Second Edition (2018), глава 4
* https://numba.pydata.org/numba-doc/latest/user/5minguide.html

## Задачи для совместного разбора

1. Сгенерируйте массив `A` из `N=1млн` случайных целых чисел на отрезке от 0 до 1000. Пусть `B[i] = A[i] + 100`. Посчитайте среднее значение массива `B`.

In [8]:
import numpy as np
import pandas as pd
from numba import jit

In [9]:
% load_ext line_profiler

The line_profiler extension is already loaded. To reload it, use:
  %reload_ext line_profiler


In [10]:
arr_A = np.random.randint(0, 1000, size=1_000_000)

In [11]:
%%timeit

sum(arr_A) / len(arr_A) + 100

62.8 ms ± 5.16 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [12]:
%%timeit

np.mean(arr_A) + 100

823 µs ± 11.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [13]:
@jit(nopython=True)
def numba_speed_up(arr):
    return sum(arr) / len(arr) + 100

%timeit numba_speed_up(arr_A)

487 µs ± 49.4 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)


2. Создайте таблицу 2млн строк и с 4 столбцами, заполненными случайными числами. Добавьте столбец `key`, которые содержит элементы из множества английских букв. Выберите из таблицы подмножество строк, для которых в столбце `key` указаны первые 5 английских букв.

In [14]:
from string import ascii_lowercase

n = 2_000_000
df = pd.DataFrame(np.random.random(size=(n, 4)), columns=[f'random_{i}' for i in range(4)])
df['key'] = np.random.default_rng().choice(list(ascii_lowercase), size=n)
df

random_0  random_1  random_2  random_3 key
0        0.449805  0.987751  0.283799  0.865578   y
1        0.365913  0.982752  0.745270  0.185312   k
2        0.003685  0.080433  0.432194  0.613571   u
3        0.716174  0.861753  0.486926  0.626057   v
4        0.109380  0.747596  0.216897  0.526058   r
...           ...       ...       ...       ...  ..
1999995  0.219007  0.154486  0.880845  0.882821   i
1999996  0.758542  0.640169  0.061719  0.883919   a
1999997  0.539231  0.282714  0.906032  0.416378   k
1999998  0.299895  0.371367  0.082990  0.381114   u
1999999  0.239335  0.643848  0.358765  0.115958   z

[2000000 rows x 5 columns]

In [15]:
def bad_solution(df_):
    letters = ['a', 'b', 'c', 'd', 'e']
    dfs = []
    for letter in letters:
        q = df_[df_['key'] == letter]
        dfs.append(q)
    return pd.concat(dfs, axis=0)

%timeit bad_solution(df)

569 ms ± 10.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [16]:
def inline_solution(df_):
    return df_[df_['key'].isin({'a', 'b', 'c', 'd', 'e'})]

%timeit inline_solution(df)

89.4 ms ± 5.81 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [17]:
def for_profiling_solution(df_):
    letters = {'a', 'b', 'c', 'd', 'e'}
    mask = df_['key'].isin(letters)
    return df_[mask]

%lprun -f for_profiling_solution for_profiling_solution(df)

## Лабораторная работа 3

1. В файлах `recipes_sample.csv` и `reviews_sample.csv` (__ЛР 2__) находится информация об рецептах блюд и отзывах на эти рецепты соответственно. Загрузите данные из файлов в виде `pd.DataFrame` с названиями `recipes` и `reviews`. Обратите внимание на корректное считывание столбца(ов) с индексами. Приведите столбцы к нужным типам.

Реализуйте несколько вариантов функции подсчета среднего значения столбца `rating` из таблицы `reviews` для отзывов, оставленных в 2010 году.

A. С использованием метода `DataFrame.iterrows` исходной таблицы;

Б. С использованием метода `DataFrame.iterrows` таблицы, в которой сохранены только отзывы за 2010 год;

В. С использованием метода `Series.mean`.

Проверьте, что результаты работы всех написанных функций корректны и совпадают. Измерьте выполнения всех написанных функций.


In [18]:
recipes_df = pd.read_csv('sem/data/recipes_sample.csv', sep=',', parse_dates=['submitted'])

recipes_df.info()
recipes_df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   name            30000 non-null  object        
 1   id              30000 non-null  int64         
 2   minutes         30000 non-null  int64         
 3   contributor_id  30000 non-null  int64         
 4   submitted       30000 non-null  datetime64[ns]
 5   n_steps         18810 non-null  float64       
 6   description     29377 non-null  object        
 7   n_ingredients   21120 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(3), object(2)
memory usage: 1.8+ MB


name      id  minutes  \
0             george s at the cove  black bean soup   44123       90   
1                healthy for them  yogurt popsicles   67664       10   
2                      i can t believe it s spinach   38798       30   
3                              italian  gut busters   35173       45   
4          love is in the air  beef fondue   sauces   84797       25   
...                                             ...     ...      ...   
29995  zurie s holey rustic olive and cheddar bread  267661       80   
29996          zwetschgenkuchen  bavarian plum cake  386977      240   
29997   zwiebelkuchen   southwest german onion cake  103312       75   
29998                                   zydeco soup  486161       60   
29999        cookies by design   cookies on a stick  298512       29   

       contributor_id  submitted  n_steps  \
0               35193 2002-10-25      NaN   
1               91970 2003-07-26      NaN   
2                1533 2002-08-29      NaN   
3               22724 2002-07-27      NaN   
4                4470 2004-02-23      4.0   
...               ...        ...      ...   
29995          200862 2007-11-25     16.0   
29996          177443 2009-08-24      NaN   
29997          161745 2004-11-03      NaN   
29998          227978 2012-08-29      NaN   
29999          506822 2008-04-15      9.0   

                                             description  n_ingredients  
0      an original recipe created by chef scott meska...           18.0  
1      my children and their friends ask for my homem...            NaN  
2                these were so go, it surprised even me.            8.0  
3      my sister-in-law made these for us at a family...            NaN  
4      i think a fondue is a very romantic casual din...            NaN  
...                                                  ...            ...  
29995  this is based on a french recipe but i changed...           10.0  
29996  this is a traditional fresh plum cake, thought...           11.0  
29997  this is a traditional late summer early fall s...            NaN  
29998  this is a delicious soup that i originally fou...            NaN  
29999  i've heard of the 'cookies by design' company,...           10.0  

[30000 rows x 8 columns]

In [19]:
reviews_df = pd.read_csv('sem/data/reviews_sample.csv', sep=',', index_col=0, parse_dates=['date'])

reviews_df.info()
reviews_df

<class 'pandas.core.frame.DataFrame'>
Int64Index: 126696 entries, 370476 to 691207
Data columns (total 5 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   user_id    126696 non-null  int64         
 1   recipe_id  126696 non-null  int64         
 2   date       126696 non-null  datetime64[ns]
 3   rating     126696 non-null  int64         
 4   review     126679 non-null  object        
dtypes: datetime64[ns](1), int64(3), object(1)
memory usage: 5.8+ MB


user_id  recipe_id       date  rating  \
370476        21752      57993 2003-05-01       5   
624300       431813     142201 2007-09-16       5   
187037       400708     252013 2008-01-10       4   
706134   2001852463     404716 2017-12-11       5   
312179        95810     129396 2008-03-14       5   
...             ...        ...        ...     ...   
1013457     1270706     335534 2009-05-17       4   
158736      2282344       8701 2012-06-03       0   
1059834      689540     222001 2008-04-08       5   
453285   2000242659     354979 2015-06-02       5   
691207       463435     415599 2010-09-30       5   

                                                    review  
370476   Last week whole sides of frozen salmon fillet ...  
624300   So simple and so tasty!  I used a yellow capsi...  
187037   Very nice breakfast HH, easy to make and yummy...  
706134   These are a favorite for the holidays and so e...  
312179   Excellent soup!  The tomato flavor is just gre...  
...                                                    ...  
1013457  This recipe was great! I made it last night. I...  
158736   This recipe is outstanding. I followed the rec...  
1059834  Well, we were not a crowd but it was a fabulou...  
453285   I have been a steak eater and dedicated BBQ gr...  
691207   Wonderful and simple to prepare seasoning blen...  

[126696 rows x 5 columns]

In [26]:
def a_solution(reviews: pd.DataFrame):
    sum_, n_ = 0, 0
    for i, row in reviews.iterrows():
        if row['date'].year == 2010:
            sum_ += row['rating']
            n_ += 1

    return sum_ / n_


%timeit a_solution(reviews_df)

4.7 s ± 379 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [28]:
def b_solution(reviews: pd.DataFrame):
    reviews = reviews[reviews['date'].dt.year == 2010]

    sum_, n_ = 0, 0
    for i, row in reviews.iterrows():
        sum_ += row['rating']
        n_ += 1

    return sum_ / n_


%timeit b_solution(reviews_df)

461 ms ± 46.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [32]:
def c_solution(reviews: pd.DataFrame):
    return reviews[reviews['date'].dt.year == 2010]['rating'].mean()


%timeit c_solution(reviews_df)

11.4 ms ± 271 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [33]:
a_result = a_solution(reviews_df)
b_result = b_solution(reviews_df)
c_result = c_solution(reviews_df)

4.4544402182900615


In [34]:
assert a_result == b_result == c_result

2. Какая из созданных функций выполняется медленнее? Что наиболее сильно влияет на скорость выполнения? Для ответа использовать профайлер `line_profiler`. Сохраните результаты работы профайлера в отдельную текстовую ячейку и прокомментируйте результаты его работы.

(*). Сможете ли вы ускорить работу функции 1Б, отказавшись от использования метода `iterrows`, но не используя метод `mean`?

In [35]:
%lprun -f a_solution a_solution(reviews_df)

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     1
     2         1         36.0     36.0      0.0
     3    126697  120483192.0    951.0     88.4
     4    126696   14671916.0    115.8     10.8
     5     12094    1115105.0     92.2      0.8
     6     12094      59278.0      4.9      0.0
     7
     8         1         13.0     13.0      0.0

In [47]:
def b_solution_speed_up(reviews: pd.DataFrame):
    def mean(x: pd.Series) -> float:
        return np.sum(x) / x.size

    reviews = reviews[reviews['date'].dt.year == 2010]
    return reviews['rating'].agg(lambda x: mean(x))


%timeit b_solution_speed_up(reviews_df)

11.9 ms ± 400 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [48]:
b_speed_up_result = b_solution_speed_up(reviews_df)

In [49]:
assert b_speed_up_result == a_result == b_result == c_result

3. Вам предлагается воспользоваться функцией, которая собирает статистику о том, сколько отзывов содержат то или иное слово. Измерьте время выполнения этой функции. Сможете ли вы найти узкие места в коде, используя профайлер? Выпишите (словами), что в имеющемся коде реализовано неоптимально. Оптимизируйте функцию и добейтесь значительного (как минимум, на один порядок) прироста в скорости выполнения.

In [ ]:
def get_word_reviews_count(df):
    word_reviews = {}
    for _, row in df.dropna(subset=['review']).iterrows():
        recipe_id, review = row['recipe_id'], row['review']
        words = review.split(' ')
        for word in words:
            if word not in word_reviews:
                word_reviews[word] = []
            word_reviews[word].append(recipe_id)

    word_reviews_count = {}
    for _, row in df.dropna(subset=['review']).iterrows():
        review = row['review']
        words = review.split(' ')
        for word in words:
            word_reviews_count[word] = len(word_reviews[word])
    return word_reviews_count

4. Напишите несколько версий функции `MAPE` (см. [MAPE](https://en.wikipedia.org/wiki/Mean_absolute_percentage_error)) для расчета среднего абсолютного процентного отклонения значения рейтинга отзыва на рецепт от среднего значения рейтинга по всем отзывам для этого рецепта. 
    1. Без использования векторизованных операций и методов массивов `numpy` и без использования `numba`
    2. Без использования векторизованных операций и методов массивов `numpy`, но с использованием `numba`
    3. С использованием векторизованных операций и методов массивов `numpy`, но без использования `numba`
    4. C использованием векторизованных операций и методов массивов `numpy` и `numba`
    
Измерьте время выполнения каждой из реализаций.

Замечание: удалите из выборки отзывы с нулевым рейтингом.


#### [версия 2]
* Уточнены формулировки задач 1, 3, 4